File: AoC_02_GiftShop.cs
var data = File.ReadAllText(@"input.txt").Split(",");
PartOne();
PartTwo();

void PartOne()
{
    long invalid = 0;
    foreach (var range in data)
    {
        string[] parts = range.Split('-');

        long begin = long.Parse(parts[0]), end = long.Parse(parts[1]);
        for (long i = begin; i <= end; ++i)
        {
            string digits = i.ToString();
            if (digits.Length % 2 != 0)
                continue;

            int j, halfLen = digits.Length / 2;
            for (j = 0; j < halfLen; ++j)
            {
                if (digits[j] != digits[j + halfLen])
                {
                    break;
                }
            }
            if (j == halfLen)
            {
                invalid += i;
            }
        }
    }

    Console.WriteLine(invalid);
}

void PartTwo()
{
    long invalid = 0;
    foreach (var range in data)
    {
        string[] parts = range.Split('-');

        long begin = long.Parse(parts[0]), end = long.Parse(parts[1]);
        for (long i = begin; i <= end; ++i)
        {
            string digits = i.ToString();
            for (int step = 1; step <= digits.Length / 2; ++step)
            {
                if (digits.Length % step != 0)
                    continue;
                bool isInvalid = true;
                for (int j = 0; j < step; ++j)
                {
                    char d = digits[j];
                    for (int k = j + step; k < digits.Length; k += step)
                        if (digits[k] != d)
                            isInvalid = false;
                }
                if (isInvalid)
                {
                    invalid += i;
                    break;
                }
            }
        }
    }

    Console.WriteLine(invalid);
}

File: AoC_02_GiftShop.zig
const std = @import("std");

pub fn main() !void {
    try part_one();
    try part_two();
}

pub fn part_one() !void {
    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    var buffer: [100]u8 = undefined;
    var file_reader = file.reader(&buffer);
    var invalid: i64 = 0;

    while (file_reader.interface.takeDelimiterExclusive(',')) |range| {
        const separator = std.mem.indexOf(u8, range, "-") orelse 0;
        const begin: i64 = try std.fmt.parseInt(i64, range[0..separator], 10);
        const end: i64 = try std.fmt.parseInt(i64, range[separator + 1 ..], 10);

        var i: i64 = begin;
        const max_len = 20;
        var buf: [max_len]u8 = undefined;
        while (i <= end) : (i += 1) {
            const digits = try std.fmt.bufPrint(&buf, "{}", .{i});
            if (@mod(digits.len, 2) != 0)
                continue;
            var j: usize = 0;
            const half: usize = digits.len / 2;
            while (j < half) : (j += 1) {
                if (digits[j] != digits[j + half])
                    break;
            }
            if (j == half) {
                invalid += i;
            }
        }
        if (!file_reader.atEnd())
            file_reader.interface.toss(1);
    } else |_| {}
    std.debug.print("{d}\n", .{invalid});
}

pub fn part_two() !void {
    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    var buffer: [100]u8 = undefined;
    var file_reader = file.reader(&buffer);
    var invalid: i64 = 0;

    while (file_reader.interface.takeDelimiterExclusive(',')) |range| {
        const separator = std.mem.indexOf(u8, range, "-") orelse 0;
        const begin: i64 = try std.fmt.parseInt(i64, range[0..separator], 10);
        const end: i64 = try std.fmt.parseInt(i64, range[separator + 1 ..], 10);

        var i: i64 = begin;
        const max_len = 20;
        var buf: [max_len]u8 = undefined;
        while (i <= end) : (i += 1) {
            const digits = try std.fmt.bufPrint(&buf, "{}", .{i});
            const half: usize = digits.len / 2;

            var step: usize = 1;
            while (step <= half) : (step += 1) {
                if (@mod(digits.len, step) != 0)
                    continue;
                var is_invalid: bool = true;
                var j: usize = 0;
                while (j < step) : (j += 1) {
                    const d: u8 = digits[j];
                    var k: usize = j + step;

                    while (k < digits.len) : (k += step) {
                        if (digits[k] != d)
                            is_invalid = false;
                    }
                }
                if (is_invalid) {
                    invalid += i;
                    break;
                }
            }
        }
        if (!file_reader.atEnd())
            file_reader.interface.toss(1);
    } else |_| {}
    std.debug.print("{d}\n", .{invalid});
}
